include/merlin/allocator.cuh

```diff
@@ -0,0 +1,97 @@
+/*
+ * Copyright (c) 2022, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include "debug.hpp"
+#include "utils.cuh"
+
+namespace nv {
+namespace merlin {
+
+enum MemoryType { Device, Managed, Pinned };
+
+class DefaultAllocator {
+ public:
+  DefaultAllocator(const DefaultAllocator&) = delete;
+  DefaultAllocator(DefaultAllocator&&) = delete;
+
+  DefaultAllocator& operator=(const DefaultAllocator&) = delete;
+  DefaultAllocator& operator=(DefaultAllocator&&) = delete;
+
+  DefaultAllocator(){};
+
+  void alloc(const MemoryType type, void** ptr, size_t size,
```

Comment:
Template the Device, Managed, Pinned case to require implemented allocator type to support difference memory type if they are used inside hkv.